src/dataflow/src/render/join.rs

```diff
@@ -375,4 +405,116 @@ where
             )
         })
     }
+
+    /// Given an input, check if it is an arrangement, and if the
+    /// arrangement can be filtered for equality to a constant.
+    /// If such filtering can be performed, render the filter as
+    /// a differential join of an arrangement to a constant collection of
+    /// size 1.
+    /// Return the RelationExpr, if any, that was rendered and the constants
+    /// in the rendered RelationExpr
+    pub fn filter_on_index_if_able(
```

Comment:
This method worries me a lot. Here are some thoughts:

1. This looks like it is starting to introduce bushy tree joins, instead of linear joins. Nothing fundamentally bad about bushy tree joins, but I don't think `render.rs` is the right place to place the brains about when and how to restructure join plans. I'd 100% like to start with linear plans for this PR if we can. If we cannot, I'd really like to have that spelled out somewhere.

2. This doesn't result in an arrangement for the result `RelationExpr` as far as I can tell, which seems odd as elsewhere (delta joins at least) the code appears to panic if we don't have the arrangement available. Or perhaps it silently makes the arrangement; I'm not sure which at the moment. But that arrangement is a potential cost for implementing the filtering this way, and isn't a clear win that could entitle `render.rs` to make that change.

Reply:
Re: 2, sorry! I'm still a bit confused because delta joins may expect multiple arrangements, by specific keys, and the `arranged_named` doesn't seem to do that.

Reply:
I guess the scarier part of 2 for me is that creating a new arrangement that the join did not ask for is a big responsibility, and we might prefer to make this clearer.